design/generic_resources.md

```diff
@@ -0,0 +1,155 @@
+
+Integer resources are for use cases where only an unsigned is needed to account
+for the resource (see Linux Realtime).
+
+A set would mostly be used for every resource which would need an
+exclusive access to it.
+
+## Constraints and Assumptions
+1. Future work might require new mechanisms to be made to allow Generic resources
+to be cluster wide in order to satisfy other use cases (e.g: pool of licenses)
+
+## Use Cases
+
+  * Exclusive access to discrete accelerators:
+    * GPU devices
+    * FPGA devices
+    * MICs (Many-Integrated Core, such as Xeon Phi)
+    * ...
+  * Support for tracking additional cgroup quotas like cpu_rt_runtime.
+    * [Linux Realtime](https://github.com/docker/docker/pull/23430)
+  * PersistentDisks in GCE
+  * Counting “slots” allowed access to a shared parallel file system.
+
+## Related Issues
+
+  * [Support abstract resource](https://github.com/docker/swarmkit/issues/594)
+  * [Add new node filter to scheduler](https://github.com/docker/swarm/issues/2223)
+  * [Add support for devices](https://github.com/docker/swarmkit/issues/1244)
+  * [Resource Control](https://github.com/docker/swarmkit/issues/211)
+  * [NVIDIA GPU support](https://github.com/docker/docker/issues/23917])
+  * [Does Docker have plan to support allocating GPU] (https://github.com/docker/docker/issues/24582)
+  * [Docker Swarm to orchestrate "Swarm Cluster" which supports GPU](https://github.com/docker/docker/issues/24750)
+  * [Use accelerator in docker container](https://github.com/docker/docker/issues/28642)
+  * [Specify resource selectors](https://github.com/docker/swarmkit/issues/206)
+
+## Objectives
+
+1. Associate multiple Generic resources with a node
+2. Request some portion of available Generic resourcess in the service
+   during service creation
+3. Enable users to define and schedule Generic resourcess in a vanilla swarmkit cluster
+
+## Non-Objectives
+
+1. Solve how Generic resourcess allocations are to be enforced or isolated.
+2. Solve how Generic resourcess are discovered
+3. Solve how cluster-level Generic resourcess should be advertised
+
+## Proposed Changes
+
+### Generic Resources request
+
+The services may only ask for Generic resourcess as integers as the solution for asking for
+specific resources can be solved in many different ways (filters, multiple kinds
+of resources, ...) and should not be addressed in this PR.
+
+```
+$ # Single resource
+$ swarmctl service create --name nginx --image nginx:latest --generic-resource "banana=2"
+$ # Multiple resources
+$ swarmctl service create --name nginx --image nginx:latest --generic-resource "banana=2;apple=3"
+```
+
+### Generic Resource advertising
+
+A node may advertise either an Integer number of resources or a set of resources.
+It is the scheduler's job to decide which resource to assign and keep track of which task
+owns which resource.
+
+```
+$ swarmd -d $DIR --join-addr $IP --join-token $TOKEN --third-party-resource "banana={blue,red,green};apple=8"
```

Comment:
ping @thaJeztah for thoughts on CLI flag syntax

While the syntax in swarmd/swarmctl isn't critical, starting the discussion will give us a head start on finding the right syntax for the Docker version. Ideally, what we merge here for swarmctl/swarmd will match what eventually goes into Docker.